File: src/unfinished_algo/kmp.cpp
#include "bits/stdc++.h"
using namespace std;

bool cmp_string(char *pat, int i, int j, int k){
    for(int x = i, y = j - k; x < i + k; x++, y++){
        if(pat[x] != pat[y]) return false;
    }

    return true;
}

vector<int> init_next_bf(char *pat){
    int m = strlen(pat);
    vector<int> nxt(m + 1, -1);

    for(int j = 1; j < m + 1; j++){
        for(int k = 0; k < j; k++){
            if(cmp_string(pat, 0, j, k)){   //Vamos fazer um hash?
                nxt[j] = k;
            }
        }
    }

    return nxt;
}

vector<int> init_next(char *pat){
    int m = strlen(pat);
    vector<int> nxt(m + 1, 0);

    nxt[0] = -1;

    int i = 1, j = 0;
    while(i + j < m){
        while(i + j < m && pat[i + j] == pat[j]){
            j++;
            next[i + j] = j;
        }

        i += (j - nxt[j]);
        j = max(0, nxt[j]);
    }

    return nxt;
}

vector<int> init_strict_next(char *pat){
    int m = strlen(pat);
    vector<int> nxt(m + 1, -1);

    if(m == 1 || (m > 1 && pat[0] != pat[1])) nxt[1] = 0;

    int i = 1, j = 0;
    while(i + j < m){
        while(i + j < m && pat[i + j] == pat[j]){
            j++;
            if(i + j == m || pat[i + j] != pat[j]) nxt[i + j] = j;
            else nxt[i + j] = nxt[j];
        }

        if(j == 0 && ((i + 1 == m) || (pat[0] != pat[i + 1]))){
            nxt[i + 1] = 0;
        }

        i += (j - nxt[j]);
        j = max(0, nxt[j]);
    }

    return nxt;
}

vector<int> kmp(char *txt, char *pat, int *nxt = NULL){
    vector<int> occ;
    int n = strlen(txt);
    int m = strlen(pat);

    if(nxt == NULL) nxt = init_next(pat);   

    int i = 0, j = 0;
    while(i <= n - m){
        while(j < m && txt[i + j] == pat[j]) j++;

        if(j == m) occ.push_back(i);
        else continue;

        i = i + (j - nxt[j]);
        j = max(0, nxt[j]);
    }

    return occ;
}

int main(){

}

File: debug/kmp.py
import sys

verbose = True

def debug(s):
    global verbose
    if verbose:
        print(s, end="") 

"""
AAAAAAAAAAAAAAAAAAAA
----      ---- 
  k        k   
--------------
      j       
"""
def init_next_bf(pat):
    m = len(pat)
    nxt = (m + 1) * [-1] 
    for j in range(1, m + 1):
        for k in range(0, j):
            if pat[:k] == pat[j-k:j]:
                nxt[j] = k
    return nxt


def init_next(pat):
    m = len(pat)
    nxt = (m+1) * [0]
    nxt[0] = -1
    nxt[1] = 0
    i, j = 1, 0
    while i + j < m:
        debug("\n"+pat+"\n")
        debug((i*" ") + (j*"="))
        while i + j < m and pat[i+j] == pat[j]:
            debug(".")
            j += 1
            nxt[i + j] = j
            #debug("match nxt[%d]=%d\n"%(i+j, j))
        if i + j < m:
            debug("!")
        debug("\n"+(i*" ")+pat+"\n")
        i += (j - nxt[j])
        j = max(0, nxt[j])
    return nxt


def init_strict_next(pat):
    m = len(pat)
    nxt = (m+1) * [-1]
    nxt[0] = -1
    if m==1 or (m > 1 and pat[0] != pat[1]):
        nxt[1] = 0
    i, j = 1, 0
    while i + j < m:
   
        while i + j < m and pat[i+j] == pat[j]:
            j += 1
            if i + j == m or pat[i + j] != pat[j]:
                nxt[i + j] = j
            else:
                nxt[i + j] = nxt[j]
            #debug("match nxt[%d]=%d\n"%(i+j, j))
        if j == 0 and ((i + 1 == m) or (pat[0] != pat[i + 1])):
            nxt[i + 1] = 0
   
        i += (j - nxt[j])
        j = max(0, nxt[j])
    return nxt



def kmp(txt, pat, nxt = None):
    occ = []
    n = len(txt)
    m = len(pat)
    nxt = init_next(pat) if not nxt else nxt
    i, j = 0, 0
    while i <= n - m:
        #debug("\n"+txt+"\n")
        #debug((i*" ") + (j*"="))
        while j < m and txt[i+j] == pat[j]:
            j += 1
            debug(".")
        if j == m:
            occ.append(i)
        else:
            #debug("!")
            pass
        #debug("\n"+(i*" ")+pat+"\n")
        i = i + (j - nxt[j])
        j = max(0, nxt[j])
    return occ


def test():
    txt = "abracadabraabraabracadabra"
    pat = "abracadabra"
    nxt = init_strict_next(pat)
    print(pat)
    print(nxt)
    occ = kmp(txt, pat, nxt)
    print(occ)    



def main():
    global verbose
    verbose = False
    pat = sys.argv[1]
    nxt = init_strict_next(pat)
    txtfile = open(sys.argv[2], "r")
    count = 0
    for line in txtfile:
        txt = line.strip("\n")
        occ = kmp(txt, pat, nxt)
        if occ:
            print(txt)
        count += len(occ)        
    txtfile.close()
    print("Total occurrences", count)


if __name__ == "__main__":
    #test()
    main()
